### PR TITLE
Replace static TagSpecs with derived AST parsed specs

### PR DESCRIPTION
**Pull request by joshuadavidthomas**
## Summary by CodeRabbit

* **New Features**
  * Template library discovery and symbol management for richer tag/filter completions.
  * Corpus tooling and CLI to add, lock, sync, and clean a real-world Django corpus.
  * Enhanced Python-side analysis to extract tag/filter rules, snippets, and block/filter arities.

* **Bug Fixes**
  * Corrected deprecation/migration references to target removal in v6.0.2.

* **Documentation**
  * Added comprehensive agent guidelines (AGENTS.md).
  * Updated site navigation to include Template Validation.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -12,7 +12,9 @@ djls-semantic = { path = "crates/djls-semantic" }
 djls-server = { path = "crates/djls-server" }
 djls-source = { path = "crates/djls-source" }
 djls-templates = { path = "crates/djls-templates" }
+djls-extraction = { path = "crates/djls-extraction" }
 djls-workspace = { path = "crates/djls-workspace" }
+djls-corpus = { path = "crates/djls-corpus" }
```
Comment (joshuadavidthomas) [inline Cargo.toml]: Alphabetize in the list.

File: Cargo.toml [hunk 2]
```diff
@@ -39,9 +45,13 @@ tracing-subscriber = { version = "0.3", features = ["env-filter", "fmt", "time"]
 walkdir = "2.5"
 which = "8.0"
 
+# python ast parsing (ruff 0.15.0)
+ruff_python_ast = { git = "https://github.com/astral-sh/ruff.git", rev = "ce5f7b6127a5d684e96fd0f8e387f73c41c7a1b0" }
+ruff_python_parser = { git = "https://github.com/astral-sh/ruff.git", rev = "ce5f7b6127a5d684e96fd0f8e387f73c41c7a1b0" }
```
Comment (joshuadavidthomas) [inline Cargo.toml]: Move to main list.

Comment (joshuadavidthomas) [inline Cargo.toml]:
```suggestion
ruff_python_ast = { git = "https://github.com/astral-sh/ruff.git", rev = "ce5f7b6127a5d684e96fd0f8e387f73c41c7a1b0" }  # 0.15.0
ruff_python_parser = { git = "https://github.com/astral-sh/ruff.git", rev = "ce5f7b6127a5d684e96fd0f8e387f73c41c7a1b0" }  # 0.15.0
```

File: crates/djls-bench/src/db.rs [hunk 3]
```diff
@@ -89,4 +88,16 @@ impl SemanticDb for Db {
...
+        None
+    }
+
+    fn filter_arity_specs(&self) -> djls_semantic::FilterAritySpecs {
+        djls_semantic::FilterAritySpecs::new()
+    }
+
+    fn environment_inventory(&self) -> Option<djls_extraction::EnvironmentInventory> {
+        None
+    }
```
Comment (joshuadavidthomas) [inline crates/djls-bench/src/db.rs]: This applies elsewhere, but this is the first appearance in the review list so I'll just leave the comment here. 

I don't like the inventory naming. I don't like the inspector naming (the inspector is just the method, not what really matters. 
I don't like that the inspector inventory is named TemplateTags and that it has more than that on there.

File: crates/djls-corpus/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+reqwest = { workspace = true }
+serde = { workspace = true }
+sha2 = { workspace = true }
+serde_json = { workspace = true }
+tar = { workspace = true }
+toml = { workspace = true }
+walkdir = { workspace = true }
+
+[dev-dependencies]
+tempfile = { workspace = true }
```
Comment (joshuadavidthomas) [inline crates/djls-corpus/Cargo.toml]: This crate is only for testing purposes so I'm not sure how meaningful this split is. Minor nitpick.

File: crates/djls-corpus/manifest.toml
Comment (joshuadavidthomas) [inline crates/djls-corpus/manifest.toml:1]: Need to add all supported versions of Wagtail (find in their docs). 

Need to add django-components, django-bird (natch), django-cotton, dj-angles, and any other component based third-party libraries. 

Need to find a way to just specify a major.minor version instead of requiring the patch version.

Need to add more actual OSS projects built on Django. Template linter prototype should have a longer list that we add.

File: crates/djls-corpus/manifest.toml [hunk 5]
```diff
@@ -0,0 +1,62 @@
+[corpus]
+root_dir = ".corpus"
```
Comment (joshuadavidthomas) [inline crates/djls-corpus/manifest.toml:2]: Not sure this really matters to have as a config option?

File: crates/djls-corpus/manifest.toml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+name = "django-compressor"
+version = "4.5.1"
+sha256 = "c1d8a48a2ee4d8b7f23c411eb9c97e2d88db18a18ba1c9e8178d5f5b8366a822"
+
+# Project repos (pinned by commit)
+[[repo]]
+name = "getsentry-sentry"
+url = "https://github.com/getsentry/sentry.git"
+ref = "f96e0f2de81f9d5b95614393be420f3107604220"
+sha256 = "9666a3f85783c84f653d565455c92efecd4e3a10a64295feaeb129a5a10b2a03"
```
Comment (joshuadavidthomas) [inline crates/djls-corpus/manifest.toml]: If we have the url, is the name really needed? If we have the git ref do we need the sha?

File: crates/djls-corpus/src/filter.rs
Comment (joshuadavidthomas) [inline crates/djls-corpus/src/filter.rs:1]: Really not sure this is necessary? At least as a dedicated file? Plus, why are we hardcoding Django tag files?

File: crates/djls-extraction/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "djls-extraction"
```
Comment (joshuadavidthomas) [inline crates/djls-extraction/Cargo.toml]: I'm just not sold on extraction as a crate name here.

File: crates/djls-python/Cargo.toml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+name = "djls-extraction"
+version = "0.0.0"
+edition = "2021"
+
+[features]
+parser = ["dep:ruff_python_parser"]
+
+[dependencies]
+ruff_python_ast = { workspace = true }
+ruff_python_parser = { workspace = true, optional = true }
```
Comment (joshuadavidthomas) [inline crates/djls-python/Cargo.toml]: I don't get gating one dep and not the other? And tbh gating it behind the feature anyway for a smallish project.

File: crates/djls-extraction/src/blocks.rs [hunk 9]
```diff
@@ -0,0 +1,1418 @@
...
+pub fn extract_block_spec(func: &StmtFunctionDef) -> Option<BlockTagSpec> {
+    let parser_var = func
+        .parameters
+        .args
+        .first()
+        .map(|p| p.parameter.name.to_string())?;
+
+    // Check for opaque block patterns first: parser.skip_past("endtag")
+    let mut skip_past_tokens = Vec::new();
+    collect_skip_past_tokens(&func.body, &parser_var, &mut skip_past_tokens);
```
Comment (joshuadavidthomas) [inline crates/djls-extraction/src/blocks.rs]: I don't like passing the vec here instead of returning? Though maybe that's idiomatic Rust?

File: crates/djls-extraction/src/blocks.rs [hunk 10]
```diff
@@ -0,0 +1,1418 @@
...
+    // Multiple parse chains via sequential control flow
+
+    #[test]
+    fn sequential_parse_then_check() {
+        // Pattern: parse → check token → conditional parse
+        let source = r#"
+def do_spaceless(parser, token):
+    nodelist = parser.parse(("endspaceless",))
+    parser.delete_first_token()
+    return SpacelessNode(nodelist)
```
Comment (joshuadavidthomas) [inline crates/djls-extraction/src/blocks.rs]: Applies to all tests here: I dunno if I like using hardcoded real templatetag functions instead of just utilizing the corpus. Reserve the hardcoded source for edge cases.

File: crates/djls-extraction/src/dataflow/domain.rs [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    Int(i64),
+    /// String constant
+    Str(String),
+    /// Tuple of tracked values (for function return/destructuring)
+    Tuple(Vec<AbstractValue>),
+}
+
+/// The abstract environment: maps variable names to their abstract values.
+#[derive(Debug, Clone, Default, PartialEq)]
+pub struct Env {
```
Comment (joshuadavidthomas) [inline crates/djls-extraction/src/dataflow/domain.rs]: I don't like the Env naming here, it's too opaque.

File: crates/djls-python/src/analysis/statements.rs
Comment (joshuadavidthomas) [inline crates/djls-python/src/analysis/statements.rs:1]: What in the world is going on with this file? Thin except for tests of other modules??

File: crates/djls-python/src/analysis/expressions.rs [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+        return AbstractValue::SplitElement {
+            index: Index::Backward(*pops_from_end + 1),
+        };
+    }
+
+    AbstractValue::Unknown
+}
+
+/// Convert an i64 to an `AbstractValue` index element based on sign.
+#[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
```
Comment (joshuadavidthomas) [inline crates/djls-python/src/analysis/expressions.rs:237]: Not great Bob

File: crates/djls-python/src/dataflow/eval/match_arms.rs [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+    }
+}
+
+/// Check if any code path in a body contains `raise TemplateSyntaxError(...)`.
+///
+/// Unlike `constraints::body_raises_template_syntax_error` (which only checks
+/// direct raises), this recurses into if/elif/else branches. Used for match
+/// case classification where any raise in any branch means the case can error.
+fn any_path_raises_template_syntax_error(body: &[Stmt]) -> bool {
+    use ruff_python_ast::StmtRaise;
```
Comment (joshuadavidthomas) [inline crates/djls-python/src/dataflow/eval/match_arms.rs]: WHY

File: crates/djls-project/src/scanning.rs [hunk 14]
```diff
@@ -0,0 +1,579 @@
...
+    tags.dedup();
+    filters.sort();
+    filters.dedup();
+
+    (tags, filters)
+}
+
+fn pathdiff(target: &Path, base: &Path) -> Option<PathBuf> {
+    target.strip_prefix(base).ok().map(PathBuf::from)
+}
```
Comment (joshuadavidthomas) [inline crates/djls-project/src/scanning.rs]: WHY

File: crates/djls-project/src/scanning.rs [hunk 15]
```diff
@@ -0,0 +1,579 @@
...
+            source_path: abs_path,
+            tags,
+            filters,
+        };
+
+        libraries.entry(load_name).or_default().push(lib);
+    }
+}
+
+fn extract_symbols_from_file(path: &Path) -> (Vec<String>, Vec<String>) {
```
Comment (joshuadavidthomas) [inline crates/djls-project/src/scanning.rs]: Why are we returning Strings instead of a richer domain object?

File: crates/djls-project/src/scanning.rs [hunk 16]
```diff
@@ -0,0 +1,579 @@
...
+        match reg.kind.symbol_kind() {
+            crate::SymbolKind::Tag => tags.push(reg.name),
+            crate::SymbolKind::Filter => filters.push(reg.name),
+        }
+    }
+
+    tags.sort();
+    tags.dedup();
+    filters.sort();
+    filters.dedup();
```
Comment (joshuadavidthomas) [inline crates/djls-project/src/scanning.rs]: Why

File: crates/djls-project/src/scanning.rs [hunk 17]
```diff
@@ -0,0 +1,579 @@
...
+        .components()
+        .map(|c| c.as_os_str().to_string_lossy())
+        .collect::<Vec<_>>()
+        .join(".")
+}
+
+fn path_to_dotted_strip_py(rel_path: &Path) -> String {
+    let dotted = path_to_dotted(rel_path);
+    dotted.strip_suffix(".py").unwrap_or(&dotted).to_string()
+}
```
Comment (joshuadavidthomas) [inline crates/djls-project/src/scanning.rs]: Why, also isn't there a path to dotted elsewhere?
